test/test_theme.rb

```diff
@@ -0,0 +1,55 @@
+require 'helper'
+
+class TestTheme < JekyllUnitTest
+  def setup
+    @theme = Theme.new('test-theme')
+    @expected_root = File.expand_path "./fixtures/test-theme", File.dirname(__FILE__)
+  end
+
+  context "initializing" do
+    should "normalize the theme name" do
+      theme = Theme.new(' Test-Theme ')
+      assert_equal "test-theme", theme.name
+    end
+
+    should "know the theme root" do
+      assert_equal @expected_root, @theme.root
+    end
+
+    should "know the theme version" do
+      assert_equal Gem::Version.new("0.1.0"), @theme.version
+    end
+
+    should "raise an error for invalid themes" do
+      assert_raises Jekyll::Errors::MissingDependencyException do
+        Theme.new("foo").version
+      end
+    end
+  end
+
+  context "path generation" do
+    [:layouts, :includes, :sass].each do |folder|
+      should "know the #{folder} path" do
+        expected = File.expand_path("_#{folder}", @expected_root)
+        assert_equal expected, @theme.public_send("#{folder}_path")
+      end
+    end
+
+    should "generate folder paths" do
+      expected = File.expand_path("./_sass", @expected_root)
+      assert_equal expected, @theme.send(:path_for, :sass)
+    end
+
+    should "not allow paths outside of the theme root" do
+      assert_equal nil, @theme.send(:path_for, "../../source")
```

Comment:
Doesn't this create `theme_dir/_../../source` because it's prefixed with the `_`? Maybe there's a way for us to get around that by doing `path_for :_layouts` and not adding the `_` when we resolve the dir. #security